src/Utilities/vector_of_constraints.jl

```diff
@@ -15,62 +15,84 @@
 # vector, it readily gives the entries of `model.constrmap` that need to be
 # updated.
 
-struct VectorOfConstraints{F<:MOI.AbstractFunction,S<:MOI.AbstractSet} <:
-       MOI.ModelLike
-    # FIXME: It is not ideal that we have `DataType` here, it might induce type
-    #        instabilities. We should change `CleverDicts` so that we can just
-    #        use `typeof(CleverDicts.index_to_key)` here.
-    constraints::CleverDicts.CleverDict{
-        MOI.ConstraintIndex{F,S},
-        Tuple{F,S},
-        typeof(CleverDicts.key_to_index),
-        typeof(CleverDicts.index_to_key),
-    }
-
-    function VectorOfConstraints{F,S}() where {F,S}
-        return new{F,S}(
-            CleverDicts.CleverDict{MOI.ConstraintIndex{F,S},Tuple{F,S}}(),
-        )
-    end
+const _VectorOfConstraintsCleverDict{F,S} = CleverDicts.CleverDict{
+    MOI.ConstraintIndex{F,S},
+    Tuple{F,S},
+    typeof(CleverDicts.key_to_index),
+    typeof(CleverDicts.index_to_key),
+}
+
+mutable struct VectorOfConstraints{
+    F<:MOI.AbstractFunction,
+    S<:MOI.AbstractSet,
+} <: MOI.ModelLike
+    # We create a lot of these objects (one for each function-set pair)! Since
+    # most users will probably only use a few of these, we can optimize our
+    # storage by using a `Union{Nothing,X}`.
+    constraints::Union{Nothing,_VectorOfConstraintsCleverDict{F,S}}
```

Comment:
What's the advantage of allowing `Nothing` here since it's already done at the level of `VectorOfConstraints` ?

Reply:
Because even if we have a non-nothing `StructOfConstraints`, it's likely we only use a few of the set types.

Reply:
You're correct. I'm getting confused. Although I think I may have tried this and it made performance worse.